Add GH constraint damping functions with options

geoffrey4444 opened the pull request:
## Proposed changes

This is the first PR of a series to change the GH constraint damping parameters from completely hard-coded functions (which must be changed any time you change the data you're evolving) to a set of functions that take input-file options. This will ultimately allow you to change options (such as the center of a gaussian) without editing the parameters by hand in the code and then recompiling.

The new set of functions are similar to PointwiseFunctions/MathFunctions, but unlike MathFunctions, they i) don't compute derivatives, ii) don't support a special interface for 1 spatial dimension, and iii) will (through a future PR) support dependence on time and on FunctionsOfTime.

The plan I'd like to suggest to implement this is the following:

1. (this PR) PR for GH damping function base class, and a gaussian + constant as an example (no time dependence yet)
2. PR updating the compute tags for GH gammas to use the damping functions from pr 1, reproducing current behavior for EvolveKerrSchild. (This would also support EvolveGaugeWave with only input-file changes, unlike now, when you would have to change the hard-coded damping parameters and recompile).
3. PR adding time dependence support, so gammas can depend on time and FunctionsOfTime, testing with a Gaussian moving at a constant velocity (e.g., for a boosted black hole)
4. PR adding triple gaussian + constant that scales Gaussian widths by the expansion factor (i.e., the damping parameters used in BBH evolutions)

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [x] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 committed:
Add GH constraint damping functions with options

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(LIBRARY GhDampingFunctions)
```
What about the name `GhConstraintDamping`? And naming the directory `ConstraintDamping` (the `Gh` part is clear because of the subdirectory nature, I think)

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,104 @@
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace GhDampingFunctions {
```
What about `GeneralizedHarmonic::ConstraintDamping`? I.e. nesting the namespaces

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,104 @@
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
```
In the cpp you'll want `Tensor.hpp` instead of `TypeAliases.hpp`

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,123 @@
+/// \endcond
+
+namespace GhDampingFunctions {
+template <size_t VolumeDim, typename Fr>
```
Sorry, it's not clear to me why you need this forward decl. Can you just remove it?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,123 @@
+class GaussianPlusConstant;
+
+/*!
+ * \ingroup GhDampingFunctionsGroup
```
If we change the namespacing to `GeneralizedHarmonic::ConstraintDamping` then we won't need any doxygen group. By documenting the `ConstraintDamping` namespace we can find the members in the documentation by going `References->Evolution Systems->Generalized Harmonic->Constraint Damping`, which I find better because it helps keep the length of the `References` section down. Does that sound okay?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,123 @@
+
+  struct Amplitude {
+    using type = double;
+    static constexpr OptionString help = {"The amplitude."};
```
how about `The amplitude of the Gaussian`?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,123 @@
+
+  struct Width {
+    using type = double;
+    static constexpr OptionString help = {"The width."};
```
how about `The width of the Gaussian`?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,123 @@
+
+  struct Center {
+    using type = std::array<double, VolumeDim>;
+    static constexpr OptionString help = {"The center."};
```
how about `The center of the Gaussian`?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,123 @@
+  void pup(PUP::er& p) override;  // NOLINT
+
+ private:
+  double constant_{};
```
I think the `double`s we usually try to initialize to `std::numeric_limits<double>::signaling_NaN()` to make bugs easier to find :)

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,123 @@
+  double amplitude_{};
+  double inverse_width_{};
+  std::array<double, VolumeDim> center_{};
+  friend bool operator==(const GaussianPlusConstant& lhs,
```
Could you move the friend function before the member variables and add a blank line between them? Not a strict rule, but I find it easier to locate member variables if friend functions aren't at the end of the class. E.g.

- friend functions
- private member functions
- member variables

If you prefer your current ordering, could you please add a blank line between the friend function and the member variables?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,63 @@
+// Distributed under the MIT License.
```
`Gh` in file name is redundant with directory names. Maybe `ConstraintDamping` would be clearer too?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,63 @@
+ * Gamma1, or Gamma2).
+ */
+template <size_t VolumeDim, typename Fr>
+class GhDampingFunction;
```
I'm again confused by this forward declaration. I don't think you need it

geoffrey4444 replied:
You're right...this was left over from when we handled the 1D case separately, but when I don't have multiple specializations, this reduced to an unneeded forward declaration. Thanks for the catch!

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,63 @@
+ * function input (i.e., the spatial coordinates) is given as a rank-1 tensor.
+ */
+template <size_t VolumeDim, typename Fr>
+class GhDampingFunction : public PUP::able {
```
We haven't been super consistent on this, but I find it easier if the base class is also in the namespace. I'd do:
```cpp
namespace GeneralizedHarmonic::ConstraintDamping {
template <...>
class DampingFunction : ...
```

nilsdeppe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/Test_GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,115 @@
+#include "Parallel/RegisterDerivedClassesWithCharm.hpp"
+#include "Utilities/Gsl.hpp"
+
+template <size_t VolumeDim, typename Fr>
```
can you `cond` the forward decls? Doxygen is sometimes a bit silly and ignores are settings and decides to search files it shouldn't...

nilsdeppe commented on tests/Unit/Helpers/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/TestHelpers.hpp:
```diff
@@ -0,0 +1,104 @@
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/Overloader.hpp"
+
+namespace TestHelpers {
```
[C++17 update] `namespace TestHelpers::GeneralizedHarmonic::ConstraintDamping`

geoffrey4444 replied:
I wonder why clang-tidy didn't flag this like it often does?

nilsdeppe commented on tests/Unit/Helpers/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/TestHelpers.hpp:
```diff
@@ -0,0 +1,104 @@
+
+  const auto member_args_tuple = std::make_tuple(member_args...);
+  const auto helper =
+      [&](const std::unique_ptr<GhDampingFunc>& gh_damping_function) noexcept {
```
can you capture things explicitly? I know it's a bit tedious, but there are edge cases where capturing everything can lead to really, really bad bugs

nilsdeppe commented on tests/Unit/Helpers/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/TestHelpers.hpp:
```diff
@@ -0,0 +1,104 @@
+ * \brief Test a GhDampingFunction by comparing to python functions
+ *
+ * The python functions must be added to TestFunctions.py in
+ * tests/Unit/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/Python.
```
note that this path and names in this comment should be updated to reflect any name changes

nilsdeppe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/GhDampingFunctions/Test_GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,115 @@
+}
+}  // namespace
+
+SPECTRE_TEST_CASE(
```
can you combine the two test cases into one by turning the test cases into functions and having a single test case call the two functions? We're trying to reduce the number of individual test cases

geoffrey4444 commented:
Thanks for the review! I just pushed a fixup

geoffrey4444 commented:
I'll try to figure the test failure tomorrow...must need a rebase or something

geoffrey4444 commented:
Rebased...the tests were failing because I needed to update OptionString -> Options::String

nilsdeppe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/Test_GaussianPlusConstant.cpp:
```diff
@@ -9,30 +9,33 @@
 #include "Parallel/RegisterDerivedClassesWithCharm.hpp"
 #include "Utilities/Gsl.hpp"
 
+// \cond
```
triple slash, not double slash (same below at endcond)

nilsdeppe reviewed (commented):
You can squash, but there are a bunch of comments I didn't mark as resolved. Did I miss them in the fixup?

geoffrey4444 commented:
Sorry for missing all those suggestions! I know you said I could squash, but it was enough I thought it would be easier if you first check over my latest fixup commit first, to make sure I didn't miss anything this time

nilsdeppe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/Test_GaussianPlusConstant.cpp:
```diff
@@ -19,15 +19,19 @@
+namespace GeneralizedHarmonic {
+namespace ConstraintDamping {
 template <size_t VolumeDim, typename Fr>
 class GhDampingFunction;
```
How about just `DampingFunction` instead of `GhDampingFunction` since it's already in the appropriate namespace?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,54 @@
+// Distributed under the MIT License.
```
Can you rename this file to just `DampingFunction.hpp` (removing `Gh` prefix)? Then the base class to `DampingFunction`, and then I think you can also get rid of the `/// \file` comment on line 4 and 5 (we only use those if what's in the file doesn't match the file name).

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,54 @@
+
+/// Holds classes implementing GhDampingFunction (functions \f$R^n \to R\f$).
+namespace GeneralizedHarmonic::ConstraintDamping {
+template <size_t VolumeDim, typename Fr>
```
can you put these forward declarations (only one so far) in a `/// \cond`? I don't want Doxygen to pick up the undocumented versions (sigh Doxygen...)

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,54 @@
+class GaussianPlusConstant;
+
+/*!
+ * Encodes a function \f$R^n \to R\f$ where n is `VolumeDim` that represents
```
There's no `brief` showing up in the documentation. Maybe just add a brief `\brief Base class defining interface for constraint damping functions.`?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,119 @@
+
+namespace GeneralizedHarmonic::ConstraintDamping {
+/*!
+ * \brief GaussianPlusConstant \f$f = C + A
```
How about `\brief "A Gaussian plus a constant:...`?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,119 @@
+      "Computes a Gaussian plus a constant about an arbitrary coordinate "
+      "center with given width and amplitude"};
+
+  WRAPPED_PUPable_decl_base_template(
```
A bit of a nitpick, but I noticed that `WRAPPED...` and the `CkMigrateMessage` constructor show up in the Doxygen (probably true in other places). It would be cleaner if we `/// \cond` them, I think.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,119 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines GeneralizedHarmonic::ConstraintDamping::GaussianPlusConstant.
```
We don't require (and prefer to not have) a `file` comment for files where the only thing implemented matches the name of the file

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,103 @@
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
```
don't use `make_with_value` as far as I can tell

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <array>
+#include <pup.h>
```
only need forward decl of `PUP::er` in header

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,103 @@
+
+#include "Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp"
+
+#include <cmath>
```
`pup.h` and `pup_stl.h`

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,119 @@
+
+#pragma once
+
+#include <array>
```
Need `cstddef`

nilsdeppe replied:
ping

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,119 @@
+#include <array>
+#include <pup.h>
+
+#include "Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp"
```
Need `DataStructures/Tensor/TypeAliases.hpp`

nilsdeppe replied:
ping

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <cstddef>
+#include <memory>
```
Does something here use `memory`? It's not clear to me what does

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,54 @@
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Utilities/MakeWithValue.hpp"
```
why `MakeWithValue`?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,54 @@
+#include <cstddef>
+#include <memory>
+
+#include "DataStructures/Tensor/Tensor.hpp"
```
Only need `TypeAliases.hpp` not `Tensor.hpp`

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GhDampingFunction.hpp:
```diff
@@ -0,0 +1,54 @@
+
+#include <cstddef>
+#include <memory>
+
```
`DataVector.hpp` (or a forward decl, I think either is fine)

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.hpp:
```diff
@@ -0,0 +1,119 @@
+           lhs.center_ == rhs.center_;
+  }
+
+  double constant_ = std::numeric_limits<double>::signaling_NaN();
```
need `limits` include for NaN

nilsdeppe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/Test_GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,121 @@
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cmath>
```
I'll leave you to double check includes here. You shouldn't need any forward decls in a test file because everything needs to be `include`d

nilsdeppe reviewed (changes requested):
Okay, rendered documentation and did Include Police (those were the last things I hadn't done). Go ahead and squash everything, including things from this :)

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/DampingFunction.hpp:
```diff
@@ -0,0 +1,56 @@
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Parallel/CharmPupable.hpp"
+
+class DataVector;
```
Need to `/// \cond` the forward decl

geoffrey4444 replied:
Sorry I missed those comments and had so many include mistakes! I'll try to catch these things next time

nilsdeppe replied:
Nothing to apologize for! I appreciate you squeezing in the time to update the code :)

nilsdeppe reviewed (commented):
There were some comments you missed last time around (my guess is GitHub collapse them on you). Please squash everything :)

geoffrey4444 commented:
Ping @nilsdeppe

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/GaussianPlusConstant.cpp:
```diff
@@ -0,0 +1,106 @@
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
```
`TypeAliases` was supposed to go in the `hpp` file :)

geoffrey4444 replied:
🤦‍♂️ just pushed and squashed that fix

markscheel commented on tests/Unit/Helpers/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/TestHelpers.hpp:
```diff
@@ -0,0 +1,111 @@
+      [&python_function_prefix, &random_value_bounds, &member_args_tuple,
+       &used_for_size](
+          const std::unique_ptr<GhDampingFunc>& gh_damping_function) noexcept {
+        // need func variable to work around GCC bug
```
Is there documentation of what this gcc bug is and why one gets around it in this way? Does it work to say `auto func = &GhDampingFunc::operator();` (so you wouldn't need to explicitly declare CallOperatorFunction)?

geoffrey4444 replied:
There is no documentation. This comment and the explicit CallOperatorFunction declaration originally appeared in #625 , by @nilsdeppe , three years ago, in ` tests/Unit/PointwiseFunctions/EquationsOfState/TestHelpers.hpp:64` in April 2018. I kept the comment when adding an analogous `tests/Unit/Helpers/PointwiseFunctions/MathFunctions/TestHelpers.hpp`, and again when adding this `TestHelpers.hpp` for constraint damping.

@nilsdeppe do you recall what the bug was? If so, maybe document in a comment on this PR?

If you both insist, I'm happing to change this to Mark's suggestion or to whatever else you prefer, but unless there's a good reason to change it, why not leave it as is for expediency's sake?

nilsdeppe replied:
I have no idea. I would try removing it and seeing if it passes the builds. Given how old the original commit is, it could very well be for a version of GCC we don't use anymore.

markscheel commented on tests/Unit/Helpers/Evolution/Systems/GeneralizedHarmonic/ConstraintDamping/TestHelpers.hpp:
```diff
@@ -0,0 +1,111 @@
+ * generally be descriptive (e.g. 'gaussian_plus_constant') of the
+ * DampingFunction.
+ *
+ * The `python_function_prefix` argument passed to `check` must be `PREFIX`. If
```
I'm not sure what this sentence means. Where is uppercase PREFIX defined?

geoffrey4444 replied:
Again, this goes back to #625, and the same comment migrated to `TestHelpers.hpp` in MathFunctions and now ConstraintDamping. I take it to mean that when you call the `check()` function, the `python_function_prefix` argument must be the same prefix used in the python file, as described in the previous paragraph. 

I'm happy to change this to something else, if you have a suggested improvement to the wording.

geoffrey4444 replied:
Remove the "PREFIX", explain that this groups the python functions...every python function must start with this prefix

markscheel reviewed (commented):
Looks good to me; just a few documentation questions.

geoffrey4444 commented:
@markscheel @nilsdeppe I removed the "gcc bug" line, and I updated the documentation as we discussed. At least one variant passed for gcc 7,8,9, so I think it's fine to no longer include that workaround

nilsdeppe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@
 # See LICENSE.txt for details.
 
 add_subdirectory(GaugeSourceFunctions)
+add_subdirectory(ConstraintDamping)
```
Alphabet

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -42,3 +42,4 @@ target_link_libraries(
   )
 
 add_subdirectory(GaugeSourceFunctions)
+add_subdirectory(ConstraintDamping)
```
Alphabet

nilsdeppe reviewed (approved):
I'm fine correcting the alphabetical issues when you add the next damping functions.
